### PR TITLE
fix for cross-builds targeting macosx

### LINKED ISSUE #1037: Compiler base options ('b_lundef', 'b_asneeded', etc) are not always valid for the cross-compiler
Have `-Wl,--no-undefined` and `-Wl,--as-needed` any meaning when cross-comliling?
In case of cross-compilation linker can't check for this correctly, I'm wrong?

I'm asking because cross-compiler to macosx ([osxcross](https://github.com/tpoechtrager/osxcross)) does not support these options, and I'm unsure is this bug of that toolchain, or bug of meson?

### PR DESCRIPTION
**Pull request by msink**
closes #1037

### CODE REVIEW

File: mesonbuild/compilers.py [hunk 1]
```diff
@@ -103,23 +103,19 @@ def is_library(fname):
...
-                                          'minsize' : [],
-                                         })
-else:
-    gnulike_buildtype_linker_args.update({'plain' : [],
-                                          'debug' : [],
-                                          'debugoptimized' : [],
-                                          'release' : ['-Wl,-O1'],
-                                          'minsize' : [],
-                                         })
+apple_buildtype_linker_args = {'plain' : [],
```
Comment (ignatenkobrain) [inline mesonbuild/compilers.py]: __[flake8]__
* [E203] whitespace before ':'

File: mesonbuild/compilers.py [hunk 2]
```diff
@@ -103,23 +103,19 @@ def is_library(fname):
...
+                               'debug' : [],
```
Comment (ignatenkobrain) [inline mesonbuild/compilers.py]: __[flake8]__
* [E203] whitespace before ':'

File: mesonbuild/compilers.py [hunk 3]
```diff
@@ -103,23 +103,19 @@ def is_library(fname):
...
+                               'debugoptimized' : [],
```
Comment (ignatenkobrain) [inline mesonbuild/compilers.py]: __[flake8]__
* [E203] whitespace before ':'

File: mesonbuild/compilers.py [hunk 4]
```diff
@@ -103,23 +103,19 @@ def is_library(fname):
...
+                               'release' : [],
```
Comment (ignatenkobrain) [inline mesonbuild/compilers.py]: __[flake8]__
* [E203] whitespace before ':'

File: mesonbuild/compilers.py [hunk 5]
```diff
@@ -103,23 +103,19 @@ def is_library(fname):
...
+                               'minsize' : [],
```
Comment (ignatenkobrain) [inline mesonbuild/compilers.py]: __[flake8]__
* [E203] whitespace before ':'

File: mesonbuild/compilers.py [hunk 6]
```diff
@@ -103,23 +103,19 @@ def is_library(fname):
...
+                               }
+
+gnulike_buildtype_linker_args = {'plain' : [],
```
Comment (ignatenkobrain) [inline mesonbuild/compilers.py]: __[flake8]__
* [E203] whitespace before ':'

File: mesonbuild/compilers.py [hunk 7]
```diff
@@ -103,23 +103,19 @@ def is_library(fname):
...
+                                 'debug' : [],
```
Comment (ignatenkobrain) [inline mesonbuild/compilers.py]: __[flake8]__
* [E203] whitespace before ':'

File: mesonbuild/compilers.py [hunk 8]
```diff
@@ -103,23 +103,19 @@ def is_library(fname):
...
+                                 'debugoptimized' : [],
```
Comment (ignatenkobrain) [inline mesonbuild/compilers.py]: __[flake8]__
* [E203] whitespace before ':'

File: mesonbuild/compilers.py [hunk 9]
```diff
@@ -103,23 +103,19 @@ def is_library(fname):
...
+                                 'release' : ['-Wl,-O1'],
```
Comment (ignatenkobrain) [inline mesonbuild/compilers.py]: __[flake8]__
* [E203] whitespace before ':'

File: mesonbuild/compilers.py [hunk 10]
```diff
@@ -103,23 +103,19 @@ def is_library(fname):
...
+                                 'minsize' : [],
```
Comment (ignatenkobrain) [inline mesonbuild/compilers.py]: __[flake8]__
* [E203] whitespace before ':'

**Comment by jpakkane** [comment]
Looks good. Just fix the space before `:` and we'll get this merged.

File: mesonbuild/compilers.py [hunk 11]
```diff
@@ -103,23 +103,19 @@ def is_library(fname):
...
-                                          'minsize': [],
-                                          })
-else:
-    gnulike_buildtype_linker_args.update({'plain': [],
-                                          'debug': [],
-                                          'debugoptimized': [],
-                                          'release': ['-Wl,-O1'],
-                                          'minsize': [],
-                                          })
+apple_buildtype_linker_args.update({'plain': [],
```
Comment (ignatenkobrain) [inline mesonbuild/compilers.py]: __[flake8]__
* [F821] undefined name 'apple_buildtype_linker_args'

File: mesonbuild/compilers.py [hunk 12]
```diff
@@ -103,23 +103,19 @@ def is_library(fname):
...
+                                    'debug': [],
+                                    'debugoptimized': [],
+                                    'release': [],
+                                    'minsize': [],
+                                    })
+
+gnulike_buildtype_linker_args.update({'plain': [],
```
Comment (ignatenkobrain) [inline mesonbuild/compilers.py]: __[flake8]__
* [F821] undefined name 'gnulike_buildtype_linker_args'

File: mesonbuild/compilers.py [hunk 13]
```diff
@@ -2128,6 +2124,8 @@ def get_buildtype_args(self, buildtype):
         return gnulike_buildtype_args[buildtype]
 
     def get_buildtype_linker_args(self, buildtype):
+        if self.gcc_type == GCC_OSX:
+            return apple_buildtype_linker_args[buildtype]
```
Comment (ignatenkobrain) [inline mesonbuild/compilers.py]: __[flake8]__
* [F821] undefined name 'apple_buildtype_linker_args'

File: mesonbuild/compilers.py [hunk 14]
```diff
@@ -2275,6 +2273,8 @@ def get_buildtype_args(self, buildtype):
         return gnulike_buildtype_args[buildtype]
 
     def get_buildtype_linker_args(self, buildtype):
+        if self.clang_type == CLANG_OSX:
+            return apple_buildtype_linker_args[buildtype]
```
Comment (ignatenkobrain) [inline mesonbuild/compilers.py]: __[flake8]__
* [F821] undefined name 'apple_buildtype_linker_args'

File: mesonbuild/compilers.py [hunk 15]
```diff
@@ -2435,6 +2435,8 @@ def get_buildtype_args(self, buildtype):
         return gnulike_buildtype_args[buildtype]
 
     def get_buildtype_linker_args(self, buildtype):
+        if mesonlib.is_osx():
+            return apple_buildtype_linker_args[buildtype]
```
Comment (ignatenkobrain) [inline mesonbuild/compilers.py]: __[flake8]__
* [F821] undefined name 'apple_buildtype_linker_args'

**Pull request merged**